src/Numeric/Sundials.hs

```diff
@@ -134,8 +136,51 @@ solve opts =
       case odeMethod opts of
         CVMethod{} -> CV.solveC
         ARKMethod{} -> ARK.solveC
-  in
-    solveCommon solveC opts
+
+    -- This is the logic to implement interruptible solver.
+    -- The solving can be interrupted at each timepoint, it reads the `ptrStop`
+    -- value to know it needs to. So actually, if the
+    -- solver is stuck in an infinite loop for any reason, there is no
+    -- solution.
+    --
+    -- The idea is as follows:
+    --   - If no exception, the `solveC` function terminates and the `wait` is
+    --   returning the result.
+    --   - If any async exception happen, the `wait` is interrupted`, the
+    --   `exception handler` set `ptrStop to 1` and the exception bubble up to
+    --   the `withAsync` which will wait for completion of the `solveC`
+    --   function. (It will also raise an async exception which will be
+    --   rethrown immediately when leaving the FFI call.
+    --   
+    solveInterruptibleWrapper consts vars env = alloca $ \ptrStop -> do
+     -- Before running the thread, we want to ensure that it won't be stopped
+     -- immediately.
+     poke ptrStop 0
+
+     -- This is the solving action
+     let solvingAction = solveC ptrStop consts vars env
+
+     -- Here for the nasty concurrency logic
+     -- We want to ensure that in case of exception, the call `poke ptrStop 1`
+     -- in order to force the solver to stop.
+     --
+     -- Hence we mask exception the time required to setup correct exception handler
+     mask $ \restore -> do
+       -- Run the computation.
+       a <- async (restore solvingAction)
+
+       -- Now we wait for the computation to be terminated.
+       -- The exception handler is set outside of the wait, so we ensure that
+       -- either we never started the async, OR the exception handler is
+       -- correctly set.
+       --
+       -- The exception handler will first ask kindly the solver to stop by
+       -- setting ptrStop to 1 AND then wait for the stop. Note that we use
+       -- uninterruptibleCancel to force the stop / wait for it in order to
+       -- ensure that no blocked thread will be left over.
+       restore (wait a) `finally` (poke ptrStop 1 >> uninterruptibleCancel a)
```

Comment:
I'm afraid about this `poke ptrStop 1` because I don't know how to ensure that the write is correctly ordonned and correctly taken into account.

However, considering that it happen after multiples RTS operations,  especially `mask`, I do hope that it will not be optimised away.

But I don't know. However, in the worse case, it will blocks like before.

Reply:
For now, let's trust goldbot:

![image](https://github.com/novadiscovery/hmatrix-sundials/assets/9705357/b325b3c9-c7f4-4e62-8600-d8bdd3b8eb99)